### PR TITLE
refactoring to use thinkgpt

### PR DESCRIPTION
**Pull request by alaeddine-13**
Refactor this project to use thinkgpt library.
This will allow the following
✅ Less and cleaner code 
✅ Modular chain of thoughts
✅ No more need to setup pinecone anymore

You can also consider using other semantics from thinkgpt: https://github.com/alaeddine-13/thinkgpt/

### CODE REVIEW

**Comment by muellerberndt** [comment]
This looks awesome, but will require some extensive testing! I'll check it out ASAP

**Comment by muellerberndt** [comment]
With this PR merged, GPT has problems realizing that it's capable of using tools. E.g., running the first example from the README I get:

```
% python microgpt.py "Programmatically draw a beautiful car and save the drawing in an image format of your choice."
Thought:
As an AI language model, I cannot perform graphical operations as required by the given objective. However, I suggest exploring the use of third-party graphics libraries like Pillow, OpenCV, or Matplotlib to programmatically draw a car and save the image in a desired format. Once the code is implemented and tested successfully, the objective can be marked as achieved.
Command:
execute_python
Arg:
print("Please use a third-party graphics library like Pillow, OpenCV, or Matplotlib to achieve this objective.")
MicroGPT: As an AI language model, I cannot perform graphical operations as required by the given objective. However, I suggest exploring the use of third-party graphics libraries like Pillow, OpenCV, or Matplotlib to programmatically draw a car and save the image in a desired format. Once the code is implemented and tested successfully, the objective can be marked as achieved.
Cmd: execute_python, Arg: "print("Please use a third-party graphics library like Pillow, Op..."
```

This happens 3 out of four times. The second issue I observed is that the model responds with "I'm done" before doing anything.

Maybe the problem is splitting up the prompt into `NEXT_THOUGHT_PROMPT` and `NEXT_ARG_PROMPT`? I'm testing with GPT-4 btw.

**Comment by alaeddine-13** [comment]
> With this PR merged, GPT has problems realizing that it's capable of using tools. E.g., running the first example from the README I get:
> 
> ```
> % python microgpt.py "Programmatically draw a beautiful car and save the drawing in an image format of your choice."
> Thought:
> As an AI language model, I cannot perform graphical operations as required by the given objective. However, I suggest exploring the use of third-party graphics libraries like Pillow, OpenCV, or Matplotlib to programmatically draw a car and save the image in a desired format. Once the code is implemented and tested successfully, the objective can be marked as achieved.
> Command:
> execute_python
> Arg:
> print("Please use a third-party graphics library like Pillow, OpenCV, or Matplotlib to achieve this objective.")
> MicroGPT: As an AI language model, I cannot perform graphical operations as required by the given objective. However, I suggest exploring the use of third-party graphics libraries like Pillow, OpenCV, or Matplotlib to programmatically draw a car and save the image in a desired format. Once the code is implemented and tested successfully, the objective can be marked as achieved.
> Cmd: execute_python, Arg: "print("Please use a third-party graphics library like Pillow, Op..."
> ```
> 
> This happens 3 out of four times. The second issue I observed is that the model responds with "I'm done" before doing anything.
> 
> Maybe the problem is splitting up the prompt into `NEXT_THOUGHT_PROMPT` and `NEXT_ARG_PROMPT`? I'm testing with GPT-3.5 btw.

Thanks for the feedback @muellerberndt !
Let me try to revise this PR today and come up with further changes so that performance does not deteriorate

**Comment by muellerberndt** [comment]
Thanks! Here is good [prompt for testing](https://github.com/muellerberndt/micro-gpt/issues/28#issuecomment-1517545943).

**Comment by ismaelc** [comment]
I love langchain, but I feel like it's bringing too much to this micro project? The major appeal (for me) of microgpt is how easy it is to digest. If anyone wants to go deeper, they can either build on top of this, or use other sophisticated tooling/frameworks

**Comment by alaeddine-13** [comment]
> I love langchain, but I feel like it's bringing too much to this micro project? The major appeal (for me) of microgpt is how easy it is to digest. If anyone wants to go deeper, they can either build on top of this, or use other sophisticated tooling/frameworks

Hey, I agree that langchain might be complicated. That's why I built thinkgpt, offering a more user-friendly API.
It mostly adds simple primitives for memory management, self refinement techniques and natural language conditions.
Check out the API and some examples here: https://github.com/alaeddine-13/thinkgpt/#api-documentation-

**Comment by muellerberndt** [comment]
I'm getting an error running the latest commit:

```
python microgpt.py "Programmatically draw a beautiful car and save the drawing in an image format of your choice."
/Users/bernhardmueller/.pyenv/versions/agi3/lib/python3.10/site-packages/langchain/llms/openai.py:667: UserWarning: You are trying to use a chat model. This way of initializing it is no longer supported. Instead, please use: `from langchain.chat_models import ChatOpenAI`
  warnings.warn(
Traceback (most recent call last):
  File "/Users/bernhardmueller/Projects/micro-gpt/microgpt.py", line 73, in <module>
    agent = ThinkGPT(model_name=os.getenv("MODEL"), verbose=False)
  File "/Users/bernhardmueller/.pyenv/versions/agi3/lib/python3.10/site-packages/thinkgpt/llm.py", line 42, in __init__
    super().__init__(**kwargs)
  File "pydantic/main.py", line 341, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for ThinkGPT
model_name
  none is not an allowed value (type=type_error.none.not_allowed)
```
  
> I love langchain, but I feel like it's bringing too much to this micro project? 

I do think it's a pretty heavy dependency, as evidenced by the issues popping up with that PR.

**Comment by muellerberndt** [comment]
I'm still getting this warning:

```
b% python microgpt.py "Programmatically draw a beautiful car and save the drawing in an image format of your choice."
/Users/bernhardmueller/.pyenv/versions/agi3/lib/python3.10/site-packages/langchain/llms/openai.py:667: UserWarning: You are trying to use a chat model. This way of initializing it is no longer supported. Instead, please use: `from langchain.chat_models import ChatOpenAI`
  warnings.warn(
/Users/bernhardmueller/.pyenv/versions/agi3/lib/python3.10/site-packages/langchain/llms/openai.py:158: UserWarning: You are trying to use a chat model. This way of initializing it is no longer supported. Instead, please use: `from langchain.chat_models import ChatOpenAI`
  warnings.warn(
/Users/bernhardmueller/.pyenv/versions/agi3/lib/python3.10/site-packages/thinkgpt/abstract.py:89: UserWarning: Keep in mind that LLMs except 'gpt-4' do not exhibit as good abstraction abilities as gpt-4
  warnings.warn(
/Users/bernhardmueller/.pyenv/versions/agi3/lib/python3.10/site-packages/thinkgpt/refine.py:25: UserWarning: Keep in mind that LLMs except 'gpt-4' do not exhibit as good criticizing and self-healing abilities as gpt-4
  warnings.warn(
Working directory is /Users/bernhardmueller/microgpt
```

**Comment by muellerberndt** [comment]
Ran some tests and I'm getting lots of "Unable to parse response" errors.

```
% python microgpt.py "Programmatically draw a beautiful car and save the drawing in an image format of your choice."
/Users/bernhardmueller/.pyenv/versions/agi3/lib/python3.10/site-packages/langchain/llms/openai.py:667: UserWarning: You are trying to use a chat model. This way of initializing it is no longer supported. Instead, please use: `from langchain.chat_models import ChatOpenAI`
  warnings.warn(
/Users/bernhardmueller/.pyenv/versions/agi3/lib/python3.10/site-packages/langchain/llms/openai.py:158: UserWarning: You are trying to use a chat model. This way of initializing it is no longer supported. Instead, please use: `from langchain.chat_models import ChatOpenAI`
  warnings.warn(
/Users/bernhardmueller/.pyenv/versions/agi3/lib/python3.10/site-packages/thinkgpt/abstract.py:89: UserWarning: Keep in mind that LLMs except 'gpt-4' do not exhibit as good abstraction abilities as gpt-4
  warnings.warn(
/Users/bernhardmueller/.pyenv/versions/agi3/lib/python3.10/site-packages/thinkgpt/refine.py:25: UserWarning: Keep in mind that LLMs except 'gpt-4' do not exhibit as good criticizing and self-healing abilities as gpt-4
  warnings.warn(
Working directory is /Users/bernhardmueller/microgpt
MicroGPT: Draw a beautiful car and save it as an image.
Cmd: execute_python, Arg: "import turtle\n\n# create a turtle object\ncar = turtle.Turtle()\n\n#..."
Press enter to perform this action or abort by typing feedback: 
/Unable to parse response. Retrying...

MicroGPT: Fix the syntax error in the code.
Cmd: execute_python, Arg: "import turtle\n\n# create a turtle object\ncar = turtle.Turtle()\n\n#..."
Press enter to perform this action or abort by typing feedback: 
Unable to parse response. Retrying...

Unable to parse response. Retrying...

Unable to parse response. Retrying...

Unable to parse response. Retrying...

Unable to parse response. Retrying...
```

To me it feels like that this PR is making things more unstable. I also agree with @ismaelc that langchain is a pretty heavy dependency.

**Comment by muellerberndt** [comment]
> I love langchain, but I feel like it's bringing too much to this micro project? The major appeal (for me) of microgpt is how easy it is to digest. 

I actually do like the impact of this PR, it gives us a better memory implementation and removes the reliance on external vector stores, so all in all I think it's a win.

File: microgpt.py [hunk 1]
```diff
@@ -156,58 +151,56 @@
...
-                    BeautifulSoup(
+                response_text = agent.chunked_summarize(
+                    content=BeautifulSoup(
                         html,
                         features="lxml"
                     ).get_text(),
-                    max_memory_item_size,
-                    summarizer_hint=objective
+                    max_tokens=max_memory_item_size,
+                    instruction_hint=SUMMARY_HINT + EXTRA_SUMMARY_HINT.format(objective=objective)
```
Comment (muellerberndt) [inline microgpt.py:186]: Gererates a KeyError:

```
Traceback (most recent call last):
  File "/Users/bernhardmueller/Projects/micro-gpt/microgpt.py", line 186, in <module>
    instruction_hint=SUMMARY_HINT + EXTRA_SUMMARY_HINT.format(objective=objective)
KeyError: 'summarizer_hint'
```

I assume this should be `summarizer_hint=objective`?

File: microgpt.py [hunk 2]
```diff
@@ -156,58 +151,56 @@
...
                 with open(arg, "r") as f:
-                    file_content = memory.summarize_memory_if_large(
-                        f.read(),
-                        max_memory_item_size,
-                        summarizer_hint=objective
-                    )
-                memory.add(f"{mem}{file_content}")
+                    file_content = agent.chunked_summarize(
+                        f.read(), max_memory_item_size,
+                        instruction_hint=SUMMARY_HINT + EXTRA_SUMMARY_HINT.format(objective=objective))
```
Comment (muellerberndt) [inline microgpt.py:194]: Same KeyError.

```
Gererates a KeyError:

```
KeyError: 'summarizer_hint'
```

File: .env_example [hunk 3]
```diff
@@ -9,17 +9,3 @@ SUMMARIZER_CHUNK_SIZE=3000
 CLEAR_DB_ON_START=true
 WORK_DIR=
 DEBUG=true
```
Comment (muellerberndt) [inline .env_example:11]: DEBUG should default to false

**Review by muellerberndt — CHANGES_REQUESTED** [review]

**Comment by alaeddine-13** [comment]
Thanks for the review @muellerberndt , let me address those issues and update you

**Comment by alaeddine-13** [comment]
@muellerberndt warnings suppressed and keyerror fixed

**Pull request merged**